analyzer/src/main/kotlin/managers/Pnpm.kt

```diff
@@ -0,0 +1,119 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.analyzer.managers
+
+import com.fasterxml.jackson.databind.JsonNode
+
+import com.vdurmont.semver4j.Requirement
+
+import java.io.File
+
+import org.ossreviewtoolkit.analyzer.AbstractPackageManagerFactory
+import org.ossreviewtoolkit.analyzer.managers.utils.NpmModuleInfo
+import org.ossreviewtoolkit.analyzer.managers.utils.hasPnpmLockFile
+import org.ossreviewtoolkit.analyzer.managers.utils.mapDefinitionFilesForPnpm
+import org.ossreviewtoolkit.model.config.AnalyzerConfiguration
+import org.ossreviewtoolkit.model.config.RepositoryConfiguration
+import org.ossreviewtoolkit.model.jsonMapper
+import org.ossreviewtoolkit.utils.common.Os
+import org.ossreviewtoolkit.utils.common.stashDirectories
+
+/**
+ * The [fast, disk space efficient package manager](https://pnpm.io/).
+ */
+class Pnpm(
+    name: String,
+    analysisRoot: File,
+    analyzerConfig: AnalyzerConfiguration,
+    repoConfig: RepositoryConfiguration
+) : Npm(name, analysisRoot, analyzerConfig, repoConfig) {
+    class Factory : AbstractPackageManagerFactory<Pnpm>("PNPM") {
+        override val globsForDefinitionFiles = listOf("package.json", "pnpm-lock.yaml")
+
+        override fun create(
+            analysisRoot: File,
+            analyzerConfig: AnalyzerConfiguration,
+            repoConfig: RepositoryConfiguration
+        ) = Pnpm(managerName, analysisRoot, analyzerConfig, repoConfig)
+    }
+
+    /**
+     * PNPM symlinks workspace modules in the `node_modules` directory, which will result in cyclic symlinks of
+     * `node_module` directories. Limit the search depth to '2' in this case as all packages are in a direct
+     * subdirectory of the `node_modules` directory, thanks to the `--shamefully-hoist` install option.
+     */
+    override val modulesSearchDepth = 2
+
+    override fun hasLockFile(projectDir: File) = hasPnpmLockFile(projectDir)
+
+    override fun processWorkspaceModuleDirs(
+        workspaceModuleDir: File,
+        scopes: Set<String>,
+        moduleDir: File
+    ): Set<NpmModuleInfo> {
+        val childNodeModulesDir = workspaceModuleDir.resolve("node_modules")
+
+        return buildSet {
+            if (childNodeModulesDir.isDirectory) {
+                // PNPM creates a separate `node_modules` directory in all workspace modules. As these dependencies
+                // are already known to the dependency tree through the `--shamefully-hoist` option, this behavior
+                // results in duplicated package entries. Stash the child `node_modules` directory to not analyze
+                // it.
+                stashDirectories(childNodeModulesDir).use {
```

Comment:
Would it make sense to move this logic into the NPM class, e.g. make the NPM class always stash the directory, without checking whether the instance is PNPM, Yarn, NPM or anything else?
So that the function does not need to be overridden?

Reply:
While I tried to implement this, I found that the PNPM implementation had an issue with the following project setup:
```
root
- package-a
  - chalk@^5.0.0
- package-b
  - chalk@^4.0.0
```
Only one `chalk` version is hoisted to the root `node_modules` directory.
Therefore, I had to remove the stashing of the child `node_modules` directories, and solve the issue with duplicated dependencies in the dependency graph by overriding `equals` in `NpmModuleInfo`.

Reply:
I also enhanced the workspaces test to test this project setup.